### PR TITLE
GitHub action optimizations

### PR DESCRIPTION
**Pull request by yashton**
Improvements to the actions:
* Uses docker images for environments to reduce time 
* ccache now functioning fixes issue #141 
* binary artifacts exported

### CODE REVIEW

**Comment by yashton** [comment]
There are outstanding changes to the build process in other pull requests, I will need to coordinate with those branches' changes.

**Comment by tangxifan** [comment]
In general, looks good to me. As you said, the pull request #152 introduces some changes in the workflow.
I will merge the pull request #152 today or tomorrow (it is also there).
Then, you just need to adapt these new dependencies in your docker image. 
I believe it will be quick to merge this PR.

File: docker/Dockerfile.base
Comment (tangxifan) [inline docker/Dockerfile.base:17]: It seems that this docker is the base for build compatibility checks, you can remove iverilog which is required by regression tests only.

File: .dockerignore
Comment (tangxifan) [inline .dockerignore:1]: This file is to guarantee that docker will not include an old version of OpenFPGA, right? Just to confirm.

Comment (yashton) [reply .dockerignore:1]: The dockerignore is used to limit which files are mounted in the docker context when building a container image. @ganeshgore introduced the original file, I'm not sure what his original plan was. I replaced the blanket ignore all with ignoring everything but the requirements.txt file so I could run pip in the build.

File: docker/Dockerfile.base
Comment (yashton) [reply docker/Dockerfile.base:17]: great, i'll remove it from the base.

**Review by srtemp — COMMENTED** [review]
Do we want to add gcc-10 in the build matrix, now that it's released?

**Comment by tangxifan** [comment]
> Do we want to add gcc-10 in the build matrix, now that it's released?

No problem for me. Up compatibility is what we want always.

**Review by srtemp — APPROVED** [review]
Everything here looks really good.  If we want gcc-10, that's kind of a separate issue. Approved.

**Comment by tangxifan** [comment]
@srtemp As there is another pull request #152 also requesting a change on the workflow script. I would like to hold this PR until that one is merged. Then this PR just need slight changes. Is that o.k.?

**Comment by yashton** [comment]
gcc-10 is not available in ubuntu 18.04. It's not a problem to update the matrix to build with a newer ubuntu version, or even mutliple versions.

**Comment by srtemp** [comment]
We could add ppa:ubuntu-toolchain-r/test for gcc10 support; not sure if that's a terrible idea though.  At some point we'll want to add different ubuntu versions, at least 20.04LTS.  
But I don't know if we need to worry about it now.

File: .dockerignore
Comment (ganeshgore) [reply .dockerignore:1]: This is important to reduce docker image size, you should add all files/directories which doesn't affect your docker image. The current content looks good

File: docker/Dockerfile.base
Comment (ganeshgore) [reply docker/Dockerfile.base:17]: It will be better if you just copy .github/workflows/install_dependency.sh in docker image and install. 
That will ensure single point change
```
COPY .github/workflows/install_dependency.sh .
RUN source install_dependency.sh
```

File: .github/workflows/build.yml
Comment (ganeshgore) [inline .github/workflows/build.yml:24]: I suggest you ignore following paths for `linux_build`
```
openfpga_flow/**
docs/**
docker/**
.github/**
```
These directory triggers unnecessary build

Comment (ganeshgore) [reply .github/workflows/build.yml:24]: I am not sure how you can skip just one of the job from `linux_build` and `regression_tests`. There is options `if` for each job. But skipping entire workflow is easier. 
I think the better implementation will be to separate `linux_build` and `regression_tests` in two different workflows.

Comment (tangxifan) [reply .github/workflows/build.yml:24]: ``openfpga_flow`` can not be ignored as we keep adding new tests
``docs`` cannot be ignored because documentation build is part of CI
``docker`` image is actually used in CI as well.
``github`` is where CI tests is located.

In the end, I do not want to see a trash corner in our repo which can bypass CI

File: .github/workflows/build.yml
Comment (ganeshgore) [inline .github/workflows/build.yml:139]: I wonder how this `ghcr.io/lnis-uofu/openfpga-buildenv-test` image will update in case there is a change in source code and regression test both.

File: .github/workflows/build.yml
Comment (ganeshgore) [reply .github/workflows/build.yml:24]: Any change in `openfpga_flow` wont require `linux_build` task to run, just `regression_test` will run.
same is true for `github`, `docker` and `doc` none of it requires testing `linux_build`.
Note: Regression test will still run in those case.

**Comment by tangxifan** [comment]
@yashton I have merged the PR #152 and there is a minor conflict on this PR, which is expected. Can you apply a fix? Also there is a Makefile now where you can compile the repository using `make compile`. Can you apply it to our CI script?

File: .github/workflows/build.yml
Comment (yashton) [reply .github/workflows/build.yml:141]: I've added in a new workflow to publish the images. It can be triggered manually from the main branch to update the images. There's no versioning right now of the images, it's always latest. If environment changes in a backwards incompatible way, that's something to add later.

**Pull request closed without merging**
